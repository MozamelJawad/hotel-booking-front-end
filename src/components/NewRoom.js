// NewRoom.js
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { createRoom, listRooms } from '../redux/rooms/roomSlice';

const NewRoom = () => {
  const cat = useSelector((state) => state.category.categories);
  const loading = useSelector((state) => state.category.loading);
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const [room, setRoom] = useState('');
  const [category, setCategory] = useState('');

  const handleRoomChange = (event) => {
    setRoom(event.target.value);
  };

  const catChange = (event) => {
    setCategory(event.target.value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    dispatch(createRoom({ name: room, category_id: category }));
    navigate('/');
    dispatch(listRooms());
  };

  if (loading) {
    return (
      <div className="d-flex justify-content-center">
        <div className="text-success spinner-border mt-5" role="status">
          <span className="visually-hidden">Loading...</span>
        </div>
      </div>
    );
  }

  return (
    <div className="col-md-8 offset-md-2 mt-5">
      <h1 className="mb-5 text-center heading">Add New Room</h1>
      <form onSubmit={(e) => handleSubmit(e)}>
        <div className="mb-3">
          <input type="text" required value={room} onChange={(e) => handleRoomChange(e)} className="form-control" placeholder="Room Name" id="room" />
        </div>
        <div className="mb-3">
          <select
            className="form-select"
            aria-label="Default select example"
            value={category}
            onChange={(e) => catChange(e)}
          >
            <option value="">Select Category</option>
            {cat.map((c) => (
              <option key={c.id} value={c.id}>
                {c.name}
              </option>
            ))}
          </select>
        </div>
        <button type="submit" className="btn btn-primary">Create</button>
      </form>
      <p className="text-center mt-3">
        <a className="btn btn-secondary" href="/">Back to Home</a>
      </p>
    </div>
  );
};

export default NewRoom;
