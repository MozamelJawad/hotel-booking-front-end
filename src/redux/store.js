// store.js

/* eslint-disable no-unused-vars */
import { configureStore } from '@reduxjs/toolkit';
// import { logger } from 'redux-logger';
import authReducer from './userAuth/authSlice';
import roomsReducer from './rooms/roomSlice';
import categoryReducer from './categories/categorySlice';

const store = configureStore({
  reducer: {
    auth: authReducer,
    rooms: roomsReducer,
    category: categoryReducer,
  },
});

export default store;
