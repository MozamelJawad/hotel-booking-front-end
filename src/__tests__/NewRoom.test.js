import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../redux/store';
import NewRoom from '../components/NewRoom';

describe('NewRoom component', () => {
  test('renders correctly', () => {
    const { getByText, getByPlaceholderText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <NewRoom />
        </MemoryRouter>
      </Provider>,
    );

    expect(getByText('Add New Room')).toBeInTheDocument();
    expect(getByPlaceholderText('Room Name')).toBeInTheDocument();
    expect(getByText('Select Category')).toBeInTheDocument();
    expect(getByText('Create')).toBeInTheDocument();
    expect(getByText('Back to Home')).toBeInTheDocument();
  });

  test('form submission works correctly', async () => {
    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <NewRoom />
        </MemoryRouter>
      </Provider>,
    );

    const roomInput = getByPlaceholderText('Room Name');
    const categorySelect = getByText('Select Category');
    const createButton = getByText('Create');

    fireEvent.change(roomInput, { target: { value: 'Test Room' } });
    fireEvent.change(categorySelect, { target: { value: '1' } });
    fireEvent.click(createButton);

    await waitFor(() => {
      expect(window.location.pathname).toBe('/');
    });
  });

  test('matches snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <NewRoom />
        </MemoryRouter>
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
