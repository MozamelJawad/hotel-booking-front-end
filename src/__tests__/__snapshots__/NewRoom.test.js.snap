// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewRoom component matches snapshot 1`] = `
<div>
  <div
    class="col-md-8 offset-md-2 mt-5"
  >
    <h1
      class="mb-5 text-center heading"
    >
      Add New Room
    </h1>
    <form>
      <div
        class="mb-3"
      >
        <input
          class="form-control"
          id="room"
          placeholder="Room Name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <select
          aria-label="Default select example"
          class="form-select"
        >
          <option
            value=""
          >
            Select Category
          </option>
        </select>
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Create
      </button>
    </form>
    <p
      class="text-center mt-3"
    >
      <a
        class="btn btn-secondary"
        href="/"
      >
        Back to Home
      </a>
    </p>
  </div>
</div>
`;
